app/code/Magento/QuoteGraphQl/Model/Resolver/AddVirtualProductsToCart.php

```diff
@@ -0,0 +1,90 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\QuoteGraphQl\Model\Resolver;
+
+use Magento\Framework\GraphQl\Config\Element\Field;
+use Magento\Framework\GraphQl\Exception\GraphQlInputException;
+use Magento\Framework\GraphQl\Query\ResolverInterface;
+use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
+use Magento\Framework\Stdlib\ArrayManager;
+use Magento\QuoteGraphQl\Model\Cart\AddProductsToCart;
+use Magento\QuoteGraphQl\Model\Cart\ExtractDataFromCart;
+use Magento\QuoteGraphQl\Model\Cart\GetCartForUser;
+
+/**
+ * Add virtual products to cart GraphQl resolver
+ *
+ * {@inheritdoc}
+ */
+class AddVirtualProductsToCart implements ResolverInterface
+{
+    /**
+     * @var ArrayManager
+     */
+    private $arrayManager;
+
+    /**
+     * @var GetCartForUser
+     */
+    private $getCartForUser;
+
+    /**
+     * @var AddProductsToCart
+     */
+    private $addProductsToCart;
+
+    /**
+     * @var ExtractDataFromCart
+     */
+    private $extractDataFromCart;
+
+    /**
+     * @param ArrayManager $arrayManager
+     * @param GetCartForUser $getCartForUser
+     * @param AddProductsToCart $addProductsToCart
+     * @param ExtractDataFromCart $extractDataFromCart
+     */
+    public function __construct(
+        ArrayManager $arrayManager,
+        GetCartForUser $getCartForUser,
+        AddProductsToCart $addProductsToCart,
+        ExtractDataFromCart $extractDataFromCart
+    ) {
+        $this->arrayManager = $arrayManager;
+        $this->getCartForUser = $getCartForUser;
+        $this->addProductsToCart = $addProductsToCart;
+        $this->extractDataFromCart = $extractDataFromCart;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
+    {
+        $cartHash = $this->arrayManager->get('input/cart_id', $args);
```

Comment:
If we are not going to use the resolver for virtual/simple products, should we add an additional check for the product type in a resolver?
Currently we can add simple product via virtual product mutation and vice versa

Reply:
We need to have 2 different resolvers with checking of product type, but all general logic should be moved to Model in a separate class

Also, need to cover functionality with API-functional tests (including negative scenarios with a wrong product type)
Thanks